File: client/src/types/tankobon-user.ts
import type { Entity } from './tankobon-entity'
import type { EntityResponse } from './tankobon-response'

export type UserEntity = Entity<UserAttributes> & {
  type: 'USER'
}

export interface UserAttributes {
  name: string
  email: string
  biography: string
  roles: UserRole[]
  createdAt: string
}

export type UserRole = 'ROLE_ADMIN' | 'ROLE_USER'
export type UserIncludes = 'avatar'
export type UserSort = 'name' | 'createdAt' | 'modifiedAt'

export type UserSuccessResponse = EntityResponse<UserEntity>

export interface EmailAvailability {
  isAvailable: boolean
}

export interface UserCreation {
  email: string
  name: string
  password: string
  roles: UserRole[]
}

export interface UserUpdate {
  id: string
  email: string
  name: string
  biography: string
  roles: UserRole[]
}

File: server/src/main/kotlin/io/github/alessandrojean/tankobon/domain/model/TankobonUser.kt
package io.github.alessandrojean.tankobon.domain.model

import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotBlank
import java.io.Serializable
import java.time.LocalDateTime
import java.util.UUID

const val ROLE_USER = "USER"
const val ROLE_ADMIN = "ADMIN"

data class TankobonUser(
  @Email(regexp = ".+@.+\\..+")
  @NotBlank
  val email: String,
  @NotBlank
  val password: String,
  val isAdmin: Boolean,
  @NotBlank
  val name: String,

  val id: String = UUID.randomUUID().toString(),
  val biography: String = "",

  override val createdAt: LocalDateTime = LocalDateTime.now(),
  override val modifiedAt: LocalDateTime = createdAt,
): Auditable, Serializable {

  @delegate:Transient
  val roles: Set<String> by lazy {
    buildSet {
      add(ROLE_USER)
      if (isAdmin) add(ROLE_ADMIN)
    }
  }

  fun canAccessLibrary(library: Library): Boolean =
    library.ownerId == id || isAdmin || library.sharedUsersIds.contains(id)
}
